common/flatpak-utils.c

```diff
@@ -6411,12 +6411,122 @@ flatpak_progress_new (FlatpakProgressCallback progress,
                                            progress_data);
 
   g_object_set_data (G_OBJECT (ostree_progress), "callback", progress);
+  g_object_set_data (G_OBJECT (ostree_progress), "callback_data", progress_data);
   g_object_set_data (G_OBJECT (ostree_progress), "last_progress", GUINT_TO_POINTER (0));
   g_object_set_data (G_OBJECT (ostree_progress), "last_total", GUINT_TO_POINTER (0));
 
   return ostree_progress;
 }
 
+static void
+copy_ostree_progress_state (OstreeAsyncProgress *source,
+                            OstreeAsyncProgress *dest)
+{
+#define COPY_OSTREE_STATE(key) G_STMT_START {                          \
+  GVariant *variant = ostree_async_progress_get_variant (source, key); \
+  if (variant)                                                         \
+    ostree_async_progress_set_variant (dest, key, variant);            \
```

Comment:
Calling `ostree_async_progress_set_variant()` multiple times will result in the progress update being made non-atomic. IIRC the existing code is careful to do all related progress updates in a single `ostree_async_progress_set()` call, so they remain grouped atomically.

We need an `OstreeAsyncProgress` method which returns the entire state of the progress,suitable for copying straight into another `OstreeAsyncProgress`. That might require API addition in upstream OSTree, and making this code be the fallback for if that’s not available.

Reply:
I think I'll rewrite this to use ostree_async_progress_get()` and `ostree_async_progress_set()` in order to only lock/unlock twice, instead of 42 times, in any case. That would seem to be a better fallback than the current code.

Reply:
Ugh, tried that and it doesn't work. `ostree_async_progress_get()` fails an assertion if the key isn't present yet, whereas `ostree_async_progress_get_variant()` just returns null in that case. I can see why they did that, otherwise they'd have to define a default value for every key somehow. So this is the best fallback we can do, I think.

Reply:
Oh, that assertion fail is even occurring with the code already in this pull request, which previously worked.

I'm not sure why it's crashing but if I move the API into libostree then it works (and is better anyway) so I think we should do that. Unfortunately I think that means we wouldn't be able to write a reliable fallback in libflatpak.